src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py

```diff
@@ -0,0 +1,357 @@
+
+        if self.force_run:
+            # cancel the previous job if it is running or pending
+            if latest_task_status in [JobStatus.RUNNING, JobStatus.PENDING]:
+                self._cancel_job(job_id)
+            return True
+
+        if latest_task['source_data_dir'] != self.source_data_dir:
+            raise PreconditionError(
+                f'There is a previous run of the same tool and load version with a different source data directory.')
+
+        if latest_task_status in [JobStatus.FAILED, JobStatus.CANCELLED]:
+            return True
+        elif latest_task_status in [JobStatus.COMPLETED, JobStatus.RUNNING, JobStatus.PENDING]:
+            raise PreconditionError(
+                f'There is a previous run of the same tool and load version that is {str(latest_task_status)}.')
+        else:
+            raise ValueError(f'Unexpected job status: {latest_task_status}')
+
+    def __init__(self, kbase_collection, load_ver, tool, source_data_dir, root_dir=loader_common_names.ROOT_DIR,
+                 force_run=False):
+        """
+        Initialize the task manager.
+
+        :param kbase_collection: KBase collection identifier name (e.g. GTDB).
+        :param load_ver: collection load version (e.g. r207.kbase.1).
+        :param tool: tool name (e.g. gtdb_tk, checkm2, etc.)
+        :param source_data_dir: source data directory.
+        :param root_dir: root directory of the collection project.
+                         Default is the ROOT_DIR defined in src/loaders/common/loader_common_names.py
+        :param force_run: if True, remove contents of the old job directory and run the job.
+        """
+
+        self.kbase_collection = kbase_collection
+        self.load_ver = load_ver
+        self.tool = tool
+        self.source_data_dir = source_data_dir
+        self.root_dir = root_dir
+        self.force_run = force_run
+
+        # job directory is named as <kbase_collection>_<load_ver>_<tool>
+        self.job_dir = os.path.join(self.root_dir, TASKFARMER_JOB_DIR,
+                                    f'{self.kbase_collection}_{self.load_ver}_{self.tool}')
+
+        self._create_job_dir(destroy_old_job_dir=self.force_run)
+        self._tasks_df = self._retrieve_all_tasks()
+
+    def submit_job(self):
+        """
+        Submit the job to slurm
+
+        Follow the steps in https://docs.nersc.gov/jobs/workflow/taskfarmer/#taskfarmer and generate all the necessary
+        files for the job submission (e.g. wrapper script, task list and batch script, etc.) before calling this function.
+        """
+
+        # check if all the required files exist
+        required_files = [tf_common.WRAPPER_FILE, tf_common.TASK_FILE, tf_common.BATCH_SCRIPT]
+        for filename in required_files:
+            if not os.path.isfile(os.path.join(self.job_dir, filename)):
+                raise ValueError(f"{filename} does not exist in {self.job_dir}")
+
+        self._check_preconditions()
```

Comment:
Hmm, it's theoretically possible that > 1 instances of the job submission script could be run at the same time for the same job dir and so > 1 sets of jobs kicked off... not sure if that's worth worrying about right now.

Reply:
I am not sure. There are a lot of edge cases that could happen during a job run. Should we worry about them at this stage? I would say no.

Reply:
Fair enough :+1: